### PR TITLE
Client cluster disconnection

### PR DESCRIPTION
**Pull request by benjaminpetit**
Fix for "Access to grain client connection status" #2551 

Now `ProxiedMessageCenter` keeps a counter of opened gateway connection. When this counter reach 0, we send an event to the client to notify that the client might have been disconnected from the cluster.

I think we still can have race condition like this:
- Client connected to GW1 (counter set to 1)
- GW1 connection lost (but no update to the counter yet)
- GW2 connection established (but no update to the counter yet)
- GW2 updates the opened connection counter (counter set to 2)
- GW1 updates the opened connection counter (counter set to 1)

But the windows seems small enough for me, so this fix seems still useful for now

### CODE REVIEW

File: test/Tester/ClientConnectionTests/ClientDisconnectionEventTests.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        public override TestCluster CreateTestCluster()
+        {
+            return new TestCluster(new TestClusterOptions(2));
+        }
+
+        [Fact, TestCategory("BVT")]
+        public async Task EventSendWhenDisconnectedFromCluster()
+        {
+            var runtime = (OutsideRuntimeClient) RuntimeClient.Current;
```
Comment (ReubenBond) [inline test/Tester/ClientConnectionTests/ClientDisconnectionEventTests.cs]: I'm trying to get rid of `RuntimeClient.Current`. Could you grab this using DI instead?

**Comment by gabikliot** [comment]
I think you are raising a very good point here @benjaminpetit .
The described window race is indeed short, but that is not the point.

First, the window may not be so short on the cluster itself. The client may see a reconnect shorty after, but from the cluster perspective it may have been longer (the period of time when there is zero silos knowing about that client may be longer then the time this client thinks he is not connected).
Second, for the type of problems we looking this callback to fix, length of disconnect may be irrelevant. If the problem is once there is a disconnect, even a short one, we loose some client specific data in the cluster, it does not matter how long it is disconnected. The data is lost.

Which means that we need to better understand what this callback is used for.
No doubt it may be usefull for some scenarios. But it may not be enough for others. I am not convinced we really need to exposue it at all. The only scenario I am aware of that needs it is the deleted persistent steam subscription problem/bug.
We discussed that at lenght in the past with @jason-bragg , and I am still of the opinion that instead of adding this callback we should simply fix the bug. Jason knows how to, we discussed that privately. I say: just fix it, and no need to expose this callback at all. 

Exposing this callback, seems to me like a violation of Orleans's principle of higher level abstractions: pushing the complexity to the client instead of solving it in the runtime. yes, our customers want to be exposed to the guts of the system, since they are used not to trust runtimes and "solve it themselves". Orleans always used to stand against that.

So to summarize: I recommend solving the problem that requires that callback and not exposing the callback at all.

**Comment by sergeybykov** [comment]
The scenarios where this problem surfaced have been primarily (my interpretation) about small deployments of 1-2-3 silos where the likelihood of a client losing connection to all of them is higher than with a larger deployment. Client stream subscriptions being deleted is one but not necessarily the only manifestation of this behavior. Even in a simple scenario of all silos getting rebooted it will take the client a while to reconnect to the gateways.

While adding an even like this does expose some of the guts of the system, in my view it is better than having a 'black box' client in such a case unable to communicate with the cluster. If later we provide better guarantee for client stream subscriptions and handle cluster reboots in a more expedient way, we can always remove the event if it's not needed anymore.

**Comment by gabikliot** [comment]
@sergeybykov , can you describe at least one scenario when this callback will be used by the client code (except the deleted  client stream subscriptions handling)?

For deleted  client stream subscriptions handling it maybe useful, but we should solve it internally in a similar way to how client code could solve it given this callback. It will be:
a) a much better solution, because we might be able to solve it better/more efficiently, since we have more access to more runtime internals.
b) we will solve it for all clients.

Except for that case, what are other usages? We we cant come with a real, requested and justified usage, we shouldn't be adding it, right?

**Comment by sergeybykov** [comment]
@gabikliot The simplest scenario I can think of is "the cluster got shut down". This is more likely to happen with small deployments of 1-2 silos. The way this situation will manifest on the client, correct me if I'm wrong, is a series of seemingly transient failures to deliver messages. An event like this would provide a more clear indication of a loss of connection to the cluster.

Similarly, if the cluster gets rebooted, depending on timing and whether or not new silos will reuse the same addresses as old ones, it will take a while for the client to rediscover the gateways to connect to. By default we requery gateways every minute. Again, not as much of a problem for larger deployments where it is less likely that all of the gateways will disappear at the same time. An event like this would 1) provide a clear indication of the situation; and 2) allow the app to reinitialize the client faster.

Today, it would also help with the cases of stream consumers/producers and observers getting disconnected. Even if and when we eliminate those possibilities, I still see value of the event because of the reasons I listed.

**Comment by gabikliot** [comment]
I think the only usage you specified is informational: client will know. It can log it, it can monitor it. But what can it do with this info to help his situation, except informational?
You wrote: reinitialize faster. How?

**Comment by sergeybykov** [comment]
>You wrote: reinitialize faster. How?

For example, it can start invoking `Initialize()` every 1-3-5 seconds. In the process it can send alerts based on the app rules, e.g. after 3 consequently failed attempts to reconnect. The important part I think is that it will have a way to explicitly execute such a logic in a non-ambiguous situation.

**Comment by gabikliot** [comment]
That wouldn't work right? Initialize will throw already initialized.

**Comment by sergeybykov** [comment]
>That wouldn't work right? Initialize will throw already initialized.

It would have to uninitialized first, you are right. Unless we change that.

**Comment by gabikliot** [comment]
The runtime can do the same: if disconnected, periodically try to reconnect. And it can do it much better: it has an access to gateway provider, so it can react faster and more intelligently. Plus of course it will solve it for all customers.

So back to my question: is there anything the client can do with this, which we cannot do better? Except for  informational, monitoring, alerting...

**Comment by jason-bragg** [comment]
> is there anything the client can do with this, which we cannot do better? Except for informational, monitoring, alerting...

I don't know.  What we know is at least two internal systems we've written need to know this, so there is reason to believe other systems built on top of Orleans may well also need this signal.  Users will build services and other features on top of the actor model and we can't anticipate their needs.

I view the disconnected signal as an information signal, and don't see any reason to hide this signal from the users.

For the immediate purposes, by exposing it publicly it allows developers of stream providers (not just the internally developed ones) to clean up whatever state they have, but also allows other advanced feature sets like Transactions, Indexing, and reactive queries (all of which are being considered) to be developed as extensions to Orleans without need of internal access.

IMO, it's a question of extensibility; not every advanced feature should be tightly integrated, requiring internal access and changes.

**Comment by gabikliot** [comment]
My recommendation is to delay exposing extensibility points, until we know for sure they are needed and cannot be solved within the runtime. Extensibility, in general, is an anti-pattern in Orleans. From day one we wanted "to lead the developer down the garden path", which means not leaving him too much choices to make mistakes. You can read about it here: https://github.com/akka/akka-meta/blob/master/ComparisonWithOrleans.md and also in this famous and Sergey's favorite picture of screws vs. car. Orleans is a car, not a set of tools. We were against turning Orleans into CORBA or WCF. What does that means? Deciding for the developer. Orleans is a **highly opinionated** model, not a general extensability framework.

I am not against extensibility per-se. I am against it without a reason. Both scenarios Sergey mentioned: fixing lost stream subscription and re-initialize, are better done by the runtime. In fact, as @benjaminpetit  and I explained, application cannot correctly fix the lost stream subscriptions (the subscription state could be lost while client not getting a "disconnected" callback at all). It should be fixed by the runtime.

Alerting/monitoring - these are valid scenarios and better provided via dedicated monitoring/alerting support. I think we know have Telemetry providers. We can hook up this signal there.

When the actual scenario from Transactions, Indexing, and reactive queries comes up, lets re-consider. 

Most importantly - we should fix the lost stream subscription bug asap (instead of providing hooks for application to try to fix it themselves), clearly showing **customer dedication** and **doing the right thing**.

**Comment by sergeybykov** [comment]
I think this particular PR isn't a good example to argue about closeness vs. extensibility. I agree with @jason-bragg:
>I view the disconnected signal as an information signal, and don't see any reason to hide this signal from the users.

It does not change the programming model or gets us any closer to CORBA/WCF. While we support other facilities like tracing and telemetry, and we could do more in the runtime, I see no reason to hide such a signal from developer.

In general, we've been drifting toward more openness and extensibility with fetures like interceptors, DI, etc., as we've learned about all kinds of scenarios people out there find for Orleans. There's still a balance to strike here, so that we don't turn Orleans into a bucket of nuts and bolts. But I don't believe this specific change brings any such danger.

**Comment by gabikliot** [comment]
OK, I disagree, but I am OK with that decision.

I still do encourage you to fix the streaming bug instead of telling clients to fix it themselves by using this signal. They will not be able to fix it correctly. And even if they could, we should fix it for everyone.

File: src/Orleans/Core/GrainClient.cs [hunk 2]
```diff
@@ -381,6 +381,15 @@ public static Streams.IStreamProvider GetStreamProvider(string name)
             return RuntimeClient.Current.CurrentStreamProviderManager.GetProvider(name) as Streams.IStreamProvider;
         }
 
+        public delegate void ConnectionToClusterLostHandler(object sender, EventArgs e);
+
+        public static event ConnectionToClusterLostHandler ClusterConnectionLost;
+
+        internal static void NotifyClusterConnectionLost()
+        {
+            ClusterConnectionLost?.Invoke(null, null);
```
Comment (gabikliot) [inline src/Orleans/Core/GrainClient.cs]: null in both?
Isn't it against the events API convention? Why using event than?

File: src/Orleans/Messaging/ProxiedMessageCenter.cs [hunk 3]
```diff
@@ -425,6 +429,19 @@ internal void UpdateClientId(GrainId clientId)
...
+        internal void OnGatewayConnectionOpen()
+        {
+            Interlocked.Increment(ref numberOfConnectedGateways);
+        }
+
+        internal void OnGatewayConnectionClosed()
+        {
+            if (Interlocked.Decrement(ref numberOfConnectedGateways) == 0)
+            {
+                GrainClient.NotifyClusterConnectionLost();
```
Comment (gabikliot) [inline src/Orleans/Messaging/ProxiedMessageCenter.cs]: should catch potential exception, log and recover. This calls into client code.

File: src/Orleans/Core/GrainClient.cs [hunk 4]
```diff
@@ -381,6 +381,15 @@ public static Streams.IStreamProvider GetStreamProvider(string name)
             return RuntimeClient.Current.CurrentStreamProviderManager.GetProvider(name) as Streams.IStreamProvider;
         }
 
+        public delegate void ConnectionToClusterLostHandler(object sender, EventArgs e);
```
Comment (gabikliot) [inline src/Orleans/Core/GrainClient.cs]: If we provide "lost" event, we will most surely need to provide "gained back" event and count. Otherwise, from pure API design perspective this looks half baked. In particular, the rule you should apply is: would it pass MSDN API review?

**Comment by sergeybykov** [comment]
>@sergeybykov , I think you jumped the gun with this review. The code still needs to be solid.

You are probably right. I was under the [false] impression the code has been reviewed enough already by others.

File: src/Orleans/Core/GrainClient.cs (hunk 4 above)
Comment (sergeybykov) [inline src/Orleans/Core/GrainClient.cs]: Not necessarily. If the recommendation is to reinitialize the client upon a connection loss, then there is no need to a "gained back" event I think.

File: src/Orleans/Messaging/ProxiedMessageCenter.cs (hunk 3 above)
Comment (jason-bragg) [inline src/Orleans/Messaging/ProxiedMessageCenter.cs]: +1
May also want to consider exposing this signal in a thread safe way.. possibly via task completion?

File: src/Orleans/Core/GrainClient.cs (hunk 2 above)
Comment (sergeybykov) [inline src/Orleans/Core/GrainClient.cs]: `sender` should conceptually be the `GrainClient` instance. Except `GrainClient` is still static. Once we make it non-static, we should send a reference to it here. It will especially make sense for the scenarios with multiple clients connecting to different clusters. The question is what to pass before then. As ugly as it looks, I think maybe `null` as a `sender` is fine in these temporary circumstances.

Is there something meaningful we'd pass as `EventArgs`?

File: src/Orleans/Messaging/ProxiedMessageCenter.cs (hunk 3 above)
Comment (sergeybykov) [inline src/Orleans/Messaging/ProxiedMessageCenter.cs]: Client is free-threaded. So it's not a concern.

File: src/Orleans/Core/GrainClient.cs (hunk 2 above)
Comment (jason-bragg) [inline src/Orleans/Core/GrainClient.cs]: If we want to pass the client when it's no longer static, and plan to do so, can we add a comment or todo noting this.  Seems like something very easy to overlook in the future.
